accounting/rewind.go

```diff
@@ -0,0 +1,188 @@
+	}
+
+	acct = account
+	var addr sdk.Address
+	addr, err = sdk.DecodeAddress(account.Address)
+	if err != nil {
+		return
+	}
+
+	// ensure that the don't attempt to rewind a special account.
+	if specialAccounts.FeeSink == addr {
+		err = MakeSpecialAccountRewindError("FeeSink")
+		return
+	}
+	if specialAccounts.RewardsPool == addr {
+		err = MakeSpecialAccountRewindError("RewardsPool")
+		return
+	}
+
+	// Get transactions and rewind account.
+	tf := idb.TransactionFilter{
+		Address:  addr[:],
+		MinRound: round + 1,
+		MaxRound: account.Round,
+	}
+	ctx2, cf := context.WithCancel(ctx)
+	// In case of a panic before the next defer, call cf() here.
+	defer cf()
+	txns, r := db.Transactions(ctx2, tf)
+	// In case of an error, make sure the context is cancelled, and the channel is cleaned up.
+	defer func() {
+		cf()
+		for range txns {
+		}
+	}()
+	if r < account.Round {
+		err = ConsistencyError{fmt.Sprintf("queried round r: %d < account.Round: %d", r, account.Round)}
+		return
+	}
+	txcount := 0
+	for txnrow := range txns {
+		if txnrow.Error != nil {
+			err = txnrow.Error
+			return
+		}
+		txcount++
+		stxn := txnrow.Txn
+		if stxn == nil {
+			return models.Account{},
+				fmt.Errorf("rewinding past inner transactions is not supported")
+		}
+		if addr == stxn.Txn.Sender {
+			acct.AmountWithoutPendingRewards += uint64(stxn.Txn.Fee)
+			acct.AmountWithoutPendingRewards -= uint64(stxn.SenderRewards)
+		}
+		switch stxn.Txn.Type {
+		case sdk.PaymentTx:
+			if addr == stxn.Txn.Sender {
+				acct.AmountWithoutPendingRewards += uint64(stxn.Txn.Amount)
+			}
+			if addr == stxn.Txn.Receiver {
+				acct.AmountWithoutPendingRewards -= uint64(stxn.Txn.Amount)
+				acct.AmountWithoutPendingRewards -= uint64(stxn.ReceiverRewards)
+			}
+			if addr == stxn.Txn.CloseRemainderTo {
+				// unwind receiving a close-to
+				acct.AmountWithoutPendingRewards -= uint64(stxn.ClosingAmount)
+				acct.AmountWithoutPendingRewards -= uint64(stxn.CloseRewards)
+			} else if !stxn.Txn.CloseRemainderTo.IsZero() {
+				// unwind sending a close-to
+				acct.AmountWithoutPendingRewards += uint64(stxn.ClosingAmount)
+			}
+		case sdk.KeyRegistrationTx:
+			// TODO: keyreg does not rewind. workaround: query for txns on an account with typeenum=2 to find previous values it was set to.
+		case sdk.AssetConfigTx:
+			if stxn.Txn.ConfigAsset == 0 {
+				// create asset, unwind the application of the value
+				assetUpdate(&acct, txnrow.AssetID, 0, stxn.Txn.AssetParams.Total)
+			}
+		case sdk.AssetTransferTx:
+			if addr == stxn.Txn.AssetSender || addr == stxn.Txn.Sender {
+				assetUpdate(&acct, uint64(stxn.Txn.XferAsset), stxn.Txn.AssetAmount+txnrow.Extra.AssetCloseAmount, 0)
+			}
+			if addr == stxn.Txn.AssetReceiver {
+				assetUpdate(&acct, uint64(stxn.Txn.XferAsset), 0, stxn.Txn.AssetAmount)
+			}
+			if addr == stxn.Txn.AssetCloseTo {
+				assetUpdate(&acct, uint64(stxn.Txn.XferAsset), 0, txnrow.Extra.AssetCloseAmount)
+			}
+		case sdk.AssetFreezeTx:
```

Comment:
We should add, `case sdk.HeartbeatTx:` assuming that's what it is called.  It should be fine to rewind past one.

Reply:
Good call: [aa75fa323d1a47fffbfd59917eda281fd240c56e](https://github.com/algorand/indexer/pull/1645/commits/aa75fa323d1a47fffbfd59917eda281fd240c56e)